### DISCUSSION TITLE
Tips for converting Allen V1 Glif network from NEST to PyGENN.

### DISCUSSION THREAD

**Discussion by russelljjarvis** [genn-team/genn#543] [General]
I wonder if anyone has experience running the Allen V1 "measured" connectome using GENN or has any ideas about the best approach to doing this?

The files that specificy cell type and wiring are shared on DropBox [here](https://www.dropbox.com/sh/w5u31m3hq6u2x5m/AABTdY1y333DbIwpzs7k0BIKa/GLIF%20Network?dl=0&subfolder_nav_tracking=1)

These connectomes are encoded in Sonata a HDF5 file format. I have divided this network into smaller populations, and I will probably need to start using [SynapseMatrixConnectivity::SPARSE](https://genn-team.github.io/genn/documentation/4/html/dd/dd5/synapseMatrixType_8h.html#aedb0946699027562bc78103a5d2a578da0459833ba9cad7cfd7bbfe10d7bbbe6e) as a container of connection matrices in GENN.

I wonder if I can build a layered network, by chaining together multiple connection matrices? My idea is that LGN (excitatory and inhibitory), and V1 (excitatory and inhibitory),  would make up four different populations, and I would use multiple connection matrices (as layers) to connect together the various cell populations.

I am aware of this work, which I believe was implemented using PyTorch tensor types: https://www.biorxiv.org/content/10.1101/2021.11.17.469025v1.full

I would like to do a similar simulated experiment but using GENN as the simulator instead.

**Comment by neworderofjamie** [comment]
We actually had a student work on implementing this model over the summer at https://github.com/williamsnider/v1_project which could be a good starting point. He got as far as implementing all the neuron models (https://github.com/williamsnider/v1_project/blob/main/compare_GLIF.ipynb) and got pretty close to parsing the model description and simulating the model (https://github.com/williamsnider/v1_project/blob/main/complete_model.py). I think the best approach (and the one he took) is to directly transfer the structure of populations to GeNN as number of populations isn't really a limitation. One thing to bear in mind is that GeNN doesn't currently support back-propagation through time (which is what Franz Scherr was using) so would have to figure out some other way to train the model

**Reply by russelljjarvis** [reply]
Thanks for all of these helpful links @neworderofjamie. Luckily our lab is uninterested in back-propagation through time, which might make things easier for me. We were aiming to use a local learning rule like STDP, and if that is too hard to do on large-scale simulation using GPU use simpler potentiation and depression rules. Probably the network won't learn as well due to unsupervised learning rules, but we are still interested in the simulation results from a network even if it learns less effectively to do an image change task.

I don't expect that a fully developed adult V1 model would be very plastic, but I would like to do simulated developmental pruning using STDP and lateral inhibition to make WTA all circuits. When I consult the Allen V1 connectome using Sonata/HDF5, it looks like all of the synaptic weights belonging to the measured network are static, so I will need to contrive in a meaningful location for the plastic synapses. The overall approach I will take will be bio-inspired but not biologically faithful, so I will probably contrive learning rules in neurons that are not supposed to be plastic.

I will investigate this MA student's work. Thanks for the leads.

**Reply by russelljjarvis** [reply]
I succeeded in getting their code to run with the CUDA backend. I am just waiting for it to complete, its a big job.

**Comment by russelljjarvis** [comment]
I forked the repository, added in some fixes to increase reproducibility, and added in a progress bar `tqdm`.
This is my fork:
https://github.com/russelljjarvis/v1_project

The problem is while still trying to build the model, the code hits some very slow pandas data frame processing. It might be faster to eschew data frames beyond this point. Maybe there is a faster way.

Progress = 0.2516% - e5Rbp4_0 to i23Htr3a_5

```
See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["dynamics_path"] = "_"
Add dynamics file
  0%|                                                                                                                                                                                   | 0/315875 [00:00<?, ?it/s]/home/rjjarvis/git/glif_net/external_project/run_simulation.py:375: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["dynamics_path"].iloc[0] = dynamics_path
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 315875/315875 [04:44<00:00, 1110.79it/s]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 315875/315875 [11:41<00:00, 450.03it/s]
complete synapse build
building v1
  0%|                                                                                                                                                                                      | 0/111 [00:00<?, ?it/s]Progress = 0.0% - e5Rbp4_0 to e5Rbp4_0
Progress = 0.0081% - e5Rbp4_0 to e5Rbp4_1
Progress = 0.0162% - e5Rbp4_0 to e5Rbp4_2
Progress = 0.0243% - e5Rbp4_0 to e5Rbp4_3
Progress = 0.0325% - e5Rbp4_0 to e5Rbp4_4


Progress = 0.0406% - e5Rbp4_0 to e5Rbp4_5
Progress = 0.0487% - e5Rbp4_0 to e5Rbp4_6
Progress = 0.0568% - e5Rbp4_0 to e5Rbp4_7
Progress = 0.0649% - e5Rbp4_0 to e5Rbp4_8
Progress = 0.073% - e5Rbp4_0 to e5Rbp4_9
Progress = 0.0812% - e5Rbp4_0 to e23Cux2_0
Progress = 0.0893% - e5Rbp4_0 to i6Pvalb_0
Progress = 0.0974% - e5Rbp4_0 to i6Pvalb_1
Progress = 0.1055% - e5Rbp4_0 to i6Pvalb_2
Progress = 0.1136% - e5Rbp4_0 to i6Pvalb_3
Progress = 0.1217% - e5Rbp4_0 to e4Scnn1a_0
Progress = 0.1299% - e5Rbp4_0 to e4Scnn1a_1
Progress = 0.138% - e5Rbp4_0 to e4Scnn1a_2
Progress = 0.1461% - e5Rbp4_0 to e4Scnn1a_3
Progress = 0.1542% - e5Rbp4_0 to e4Scnn1a_4
Progress = 0.1623% - e5Rbp4_0 to e4Scnn1a_5
Progress = 0.1704% - e5Rbp4_0 to i23Pvalb_0
Progress = 0.1786% - e5Rbp4_0 to i23Pvalb_1
Progress = 0.1867% - e5Rbp4_0 to i23Pvalb_2
Progress = 0.1948% - e5Rbp4_0 to i6Htr3a_0
Progress = 0.2029% - e5Rbp4_0 to i6Htr3a_1
Progress = 0.211% - e5Rbp4_0 to i23Htr3a_0
Progress = 0.2191% - e5Rbp4_0 to i23Htr3a_1
Progress = 0.2273% - e5Rbp4_0 to i23Htr3a_2
Progress = 0.2354% - e5Rbp4_0 to i23Htr3a_3
Progress = 0.2435% - e5Rbp4_0 to i23Htr3a_4
Progress = 0.2516% - e5Rbp4_0 to i23Htr3a_5
```

**Reply by neworderofjamie** [reply]
Awesome - glad you got it running! I feel the data-loading could be optimised somewhat (although, I think that code saves a pre-processed pickled version once it's done all the panda-wrangling) but, the **best** solution would be to convert the connectivity rules to GeNN and generate all the connectivity on device. https://genn-team.github.io/genn/documentation/4/html/d5/dd4/sectSparseConnectivityInitialisation.html describes the basics of the system but some of the 'snippet' definitions in https://github.com/genn-team/genn/blob/master/include/genn/genn/initSparseConnectivitySnippet.h might give you a better idea of how you can generate more typical random connectivity for models like these.

**Reply by russelljjarvis** [reply]
> ...  but, the best solution would be to convert the connectivity rules to GeNN and generate all the connectivity on device. https://genn-team.github.io/genn/documentation/4/html/d5/dd4/sectSparseConnectivityInitialisation.html describes the basics of the system but some of the 'snippet' definitions in https://github.com/genn-team/genn/blob/master/include/genn/genn/initSparseConnectivitySnippet.h might give you a better idea of how you can generate more typical random connectivity for models like these.

AOK thanks @neworderofjamie, I will look into this today, perhaps I can make progress with this.

**Reply by russelljjarvis** [reply]
Hi @williamsnider,

I am really interested in reproducing aspects of your GSoC `v1_project`. So much so that I forked your repo, and got it to run on my laptop.

It is taking a long time for me to do the data frame processing to generate files used by `GeNN/GeNN-model`

I wonder if you have access to a free file hosting like cloudstor or an institution's paid DropBox, and if so would you be able to share pickled files corresponding to the processed data frame, just before GeNN model is instantiated? If so that would be awesome. I will make sure to cite any pre-pubs or pubs associated with your GSoC work.

As an example, I have used `cloudstor` via my University to host some processed Allen Brain Sonata HDF5 files that I call in code.

here: https://github.com/russelljjarvis/getConnectomes/blob/main/load_allen_measured_model.jl#L9
This downloads a directory, or at least I hope it does.

**Reply by russelljjarvis** [reply]
Hi @williamsnider, I was also wondering how to change the file[https://github.com/williamsnider/v1_project/blob/main/complete_model.py](complete_model.py) so that it executes LIF cell models, instead of GLIF cell models..

I may have answered my own question, it looks like I would change. An argument to a method `construct_populations` defined in utilities.

https://github.com/williamsnider/v1_project/blob/main/utilities.py#L11-L134

**Reply by williamsnider** [reply]
Hi @russelljjarvis, I'm glad you're using the repo! I have been working on an improved script [build_model.py](https://github.com/williamsnider/v1_project/blob/main/build_model.py) to build the model V1 model, which is a work in progress. Specifically, right now it just has the V1 nodes and V1 to V1 edges / synapses populations. It currently builds when I use the "SingleThreadedCPU" backend, and I'm trying to optimize it to build with the "CUDA" backend without running out of RAM. For your purposes this script may be useful since it is a lot clearer than the complete_model.py and also reduces the size of the pandas dataframes so the wrangling is faster. To answer your latter question, "GLIF_models.py" contains the 5 different Allen GLIF models, where GLIF1 is LIF. If you import GLIF1 you can use it when you do model.add_neuron_population(...). Hope this helps

**Reply by russelljjarvis** [reply]
Hi @williamsnider, this is awesome. Thanks for all of this.

I wonder if I should PR back any changes? It would be cool to have something that checks if the Allen/Sonata files already exists, and if they don't they get downloaded. 

Another minor thing that could help a little is `tqdm` progress bars, as these can give you running average loop completion time estimates.

I really appreciate all of the work you are doing! Thanks :)

**Reply by russelljjarvis** [reply]
In the printed progress, I put a newline character `\n`, as with the other character, the word was being overwritten, and without those key words, I had no idea in the code I was waiting for progress.

For example "Progress = {}%", only printed once but everything to the right of it was visible because of refreshing, [I did it like this](https://github.com/russelljjarvis/v1_project/blob/main/run_simulation.py#L352-L356
)

```python
        print(
            "Progress = {}% - {} to {}".format(
                np.round(100 * count / len(v1_model_names) ** 2, 4), pop1, pop2
            ),
            end="\n",
        )
```
Also, the parallel map is compatible with progress [bars via a hack](https://github.com/russelljjarvis/v1_project/blob/main/run_simulation.py#L410-L415
)


I am not running this code, on HPC, so I set the max number of workers to a lower bound, at unlimited number of workers the code fails on my machine:

```python
try:
    with multiprocessing.Pool(8, maxtasksperchild=8) as pool:
        list(tqdm(pool.imap_unordered(make_src_tgt_df, items), total=len(items)))
except:
    list(tqdm(map(make_src_tgt_df, items), total=len(items)))
```

Finally, I used the pip install tool black to format the code to make it more congruent with PEP8 with minimal effort, I am only noting this in case it looks like I wasted a lot of time making arbitrary format changes, and because the code has not changed in substance, its only cosmetically different.

**Comment by russelljjarvis** [comment]
I agree build_model.py is way faster, thanks!

**Reply by russelljjarvis** [reply]
I now have this directory full of GeNN code:
```
~/git/v1_project2/GeNNModel_CODE$
```
* Could I read I use this file to get the same connection matrix used by the GeNN model into a different language like Julia?
```
GeNNModel_CODE/synapseUpdate.cc
```
Assuming I also provided the headers to the cpp call? 
```
GeNNModel_CODE/definitionsInternal.h"
GeNNModel_CODE/supportCode.h"
```
~/git/v1_project2/GeNNModel_CODE$


The model seems to be able to partially build but I think it ultimately hits a memory failure on my CUDA laptop (32GB of system RAM, GPU memory unknown). 
```
gcc: fatal error: Killed signal terminated program cc1plus
compilation terminated.
make: *** [Makefile:22: runner.o] Error 1
make: Leaving directory '/home/rjjarvis/git/v1_project2/GeNNModel_CODE'
Traceback (most recent call last):
  File "/home/rjjarvis/git/v1_project2/build_model.py", line 272, in <module>
    model.build(force_rebuild=False)
  File "/home/rjjarvis/git/GeNN/pygenn/genn_model.py", line 616, in build
    check_call(["make", "-j", str(cpu_count(logical=False)), "-C", output_path])
  File "/home/rjjarvis/miniconda3/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['make', '-j', '8', '-C', './GeNNModel_CODE']' returned non-zero exit status 2.
```
* I wonder if I would be able to write the output spiketrains to file without necessarily loading them back into Python?

**Reply by neworderofjamie** [reply]
The connection matrices aren't in the GeNN code, they're loaded at runtime from the Python data.

**Comment by russelljjarvis** [comment]
Changing the jobs argument to 1, I am still getting the same error. 
(base) rjjarvis@rjjarvis:~/git/v1_project2$ make -j 1 -C ./GeNNModel_CODE/
make: Entering directory '~/git/v1_project2/GeNNModel_CODE'


gcc: fatal error: Killed signal terminated program cc1plus
compilation terminated.
make: *** [Makefile:22: runner.o] Error 1
make: Leaving directory '/home/rjjarvis/git/v1_project2/GeNNModel_CODE'


I checked dmesg and it looks like I ran out of some kind of memory. Do you think that decreasing the model size even more will free up memory?

[29360.557086] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=/,mems_allowed=0,global_oom,task_memcg=/,task=cc1plus,pid=21827,uid=1000
[29360.557092] Out of memory: Killed process 21827 (cc1plus) total-vm:13003196kB, anon-rss:12309020kB, file-rss:0kB, shmem-rss:0kB, UID:1000 pgtables:24244kB oom_score_adj:0
[29360.891762] oom_reaper: reaped process 21827 (cc1plus), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB

**Reply by neworderofjamie** [reply]
This is a problem with the compiler running out of memory - you **could** try on a machine with 64gb of memory but really it needs further investigation on our side.

**Reply by russelljjarvis** [reply]
Cool. I have simply switched to SingleThreadedCPU backend on the smaller subset of V1 only model. If the number of simulated time steps is an integer, this model evaluates, which is good! I will try to plot the resulting raster plot.

For short run times the model outputs no spikes, as evidenced by:

This is my different version of build_model.py 
https://gist.github.com/russelljjarvis/8850a29fb317d6e22fa0918a340f1ade


```python
print([v.spike_event_times for v in pop_dict.values()])
```
outputs a list of Nones.

I think this is because the file build_model.py doesn't define netstimulation artificial inputs it is just a prototype for building the model, for executing the model, I would need to combine code from the other slower run_model.py file like this:
```python
    pop_dict[lgn_model_name].set_extra_global_param("spikeTimes", spike_times)
    print("Added {}".format(lgn_model_name))
```

Currently, I have glued together the LGN code that receives spikes stored from disk, with the current approach in build_model.py that builds a small network more quickly. 

I think I am getting an error about badly trying to connect together populations of different sizes (LGN versus subsampled V1). This is the error message I am getting:
```stacktrace
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File ~/git/v1_project2/build_model.py:345, in <module>
    341 pdb.set_trace()
    343 ### Run simulation
--> 345 import time
    347 start = time.time()
    348 model.build(force_rebuild=False)

File ~/git/GeNN/pygenn/genn_model.py:664, in GeNNModel.load(self, path_to_model, num_recording_timesteps)
    662 # Loop through neuron populations
    663 for pop_data in itervalues(self.neuron_populations):
--> 664     pop_data.load(num_recording_timesteps)
    666 # Loop through synapse populations
    667 for pop_data in itervalues(self.synapse_populations):

File ~/git/GeNN/pygenn/genn_groups.py:587, in NeuronGroup.load(self, num_recording_timesteps)
    583     self.spike_que_ptr = self._model._slm.assign_external_pointer_single_ui(
    584         "spkQuePtr" + self.name)
    586 # Load neuron state variables
--> 587 self._load_vars(self.neuron.get_vars())
    589 # Load neuron extra global params
    590 self._load_egp()

File ~/git/GeNN/pygenn/genn_groups.py:263, in Group._load_vars(self, vars, size, var_dict, get_location_fn)
    261     # If manual initialisation is required, copy over variables
    262     if var_data.init_required:
--> 263         var_data.view[:] = var_data.values
    264 else:
    265     assert not var_data.init_required

ValueError: could not broadcast input array from shape (17383,) into shape (2250,)
```

Has anyone had luck running a scaled-down version of the NEST model?

**Reply by neworderofjamie** [reply]
Looks like one of the neuron variables you're initialising with an array/list isn't the correct size. I'm guessing this could be an issue with the ``spikes_list_to_start_end_times`` function but I also don't fully understand how the code to setup the GLIF neuron population state variables works so I guess that could also be initialising variables with arrays. We could do with better exception handling here but, adding a ``print(self.name)`` at genn_groups.py:587 would probably find the culprit.

Couple of other things:
- Why all the ``tolist()`` calls? GeNN uses numpy internally and I can only imagine lists of  weights are going to extremely large
- ``force_rebuild=True`` is really going to slow down your iteration time on a model of this complexity! It also masks bugs with models not being rebuilt so we can't fix them!

**Reply by russelljjarvis** [reply]
Cool, thanks for the pointers!
* There is not much reason for my syntax, I was just doing trial-and-error experiments using code chunks I borrowed from the other files discussed here.
* I will try instead with these recommendations.

**Comment by neworderofjamie** [comment]
In https://github.com/genn-team/genn/pull/554 I've just implemented a workaround for the slow build times/huge memory usage. Have a go on that branch, passing ``generateSimpleErrorHandling=True`` through to the ``GeNNModel`` constructor and let me know how you get on.

**Comment by russelljjarvis** [comment]
Okay I merged in the [compiler_friendly_error_handling](https://github.com/genn-team/genn/tree/compiler_friendly_error_handling) branch onto my local GeNN repository. I am rerunning the model with that GeNN installation patched via `compiler_friendly_error_handing`

**Reply by russelljjarvis** [reply]
I am still getting this:
```
sON_TF8_0
```
```python
Traceback (most recent call last):
  File "/home/rjjarvis/git/v1_project2/snider.py", line 674, in <module>
    model.load(
  File "/home/rjjarvis/git/GeNN/pygenn/genn_model.py", line 664, in load
    pop_data.load(num_recording_timesteps)
  File "/home/rjjarvis/git/GeNN/pygenn/genn_groups.py", line 591, in load
    self._load_vars(self.neuron.get_vars())
  File "/home/rjjarvis/git/GeNN/pygenn/genn_groups.py", line 263, in _load_vars
    var_data.view[:] = var_data.values
ValueError: could not broadcast input array from shape (17383,) into shape (2250,)
```

**Reply by neworderofjamie** [reply]
did you try placing a print/using a debugger where I suggested to figure out which population was causing the issue? The fix was only for compile times with the CUDA backend

**Reply by russelljjarvis** [reply]
Sorry yes, I did do that. The last printout from that line of code is `sON_TF8_0`, which comes from the spike array dedicated for managing the retinal input.

**Comment by neworderofjamie** [comment]
@russelljjarvis I have done a bit more work on this. https://github.com/neworderofjamie/allen_v1 can now simulate the full V1 model (I haven't fully checked the output but it **looks** good) using [this](https://github.com/genn-team/genn/tree/constant_cache_usage_estimate) branch of GeNN. Performance is pretty good (0.6s to simulate 1s of realtime on my workstation) although load time is still a few minutes.

**Reply by russelljjarvis** [reply]
Wow, this is great work and kind of unexpected (unexpected because of a big data pre-processing job it necessitated). 

I will try it out ASAP and let you know how it goes.

**Reply by neworderofjamie** [reply]
No worries, it was a fun pandas-learning exercise and it seemed a shame to leave this project hslf done. All you need to do is download the v1_point, lgn_stimulus and bkg_inputs folders from https://www.dropbox.com/sh/w5u31m3hq6u2x5m/AAAlNsSMiVOfg204WsPSG707a/simulations into the same folder.

**Reply by russelljjarvis** [reply]
The branch link for GeNN is broken.

**Reply by russelljjarvis** [reply]
But its clear you must mean this branch
https://github.com/genn-team/genn/tree/constant_cache_usage_estimate_2

**Reply by neworderofjamie** [reply]
Ahh, sorry about that. I changed my mind on one of modifications I made to GeNN and deleted the old branches. That is indeed the correct branch to use. If **that** gets deleted it will be because it's been merged into master!

**Reply by russelljjarvis** [reply]
Great, no worries. 

I like it how you have made an awesome (sort of inadvertent) replacement Sonata/HDF5 reader that is fast, and flexible.

I think one reason of many reasons your code is fast is because it retains the low-precision UInt type of the cell indexs, and you defer casting the HDF5 generator to a list until you populate the data frame with it (this is just speculation, I am not 100% sure why the pre-processing job is so much faster now). 

I am not sure if it is worth trying the model on just any Nvidia GPU, should jump to a HPC  Nvidia GPU, which will likely have more on board RAM.


I will flag someone else in my lab who should be aware of this new approach to simulating the Billeh V1 model.

@pabloabur

**Reply by neworderofjamie** [reply]
Hah, it might have been better to learn how to use pySonata but doing it myself wasn't too hard. Keeping everything in pandas i.e. C++ is definitely the reason why it's much faster - anything going through Python lists will be too slow and too memory-hungry.

You should be fine on any GPU really - this does need around 32GB of RAM to load but it only actually 2GB of GPU memory.

**Reply by russelljjarvis** [reply]
> Hah, it might have been better to learn how to use pySonata but doing it myself wasn't too hard. Keeping everything in pandas i.e. C++ is definitely the reason why it's much faster - anything going through Python lists will be too slow and too memory-hungry.
> 
> You should be fine on any GPU really - this does need around 32GB of RAM to load but it only actually 2GB of GPU memory.

I think the branch must be merged into master now, as I can no longer find the branch.

**Comment by russelljjarvis** [comment]
I got pretty close.

It built the model.

There is a populated directory called:

 v1_point_CODE
 ```
(base) rjjarvis@pop-os:~/git/jamie$ cd v1_point_CODE/
(base) rjjarvis@pop-os:~/git/jamie/v1_point_CODE$ ls
customUpdateCUDAOptim.cc  definitionsInternalCUDAOptim.h  neuronUpdateCUDAOptim.cc  supportCodeCUDAOptim.h
definitionsCUDAOptim.h    initCUDAOptim.cc                runnerCUDAOptim.cc        synapseUpdateCUDAOptim.cc
```


~/git/jamie$ python run.py 
Reading SONATA model description
	Nodes
		v1
			230924 neurons in 111 homogeneous populations
		lgn
			17400 neurons in 14 homogeneous populations
		bkg
			1 neurons in 1 homogeneous populations
		0.04785227999673225 seconds
	Inputs
		0.00026435799372848123 seconds
	Edges
		v1_to_v1 (v1->v1)
			Group 0
				70139111 edges in 9818 homogeneous populations
		lgn_to_v1 (lgn->v1)
			Group 0
				3506881 edges in 1195 homogeneous populations
		bkg_to_v1 (bkg->v1)
			Group 0
				230924 edges in 111 homogeneous populations
		34.170160175010096 seconds
Creating GeNN model
	Creating GeNN neuron populations
		0.07666084599622991 seconds
	Creating GeNN synapse populations
		6.34699087899935 seconds
Building GeNN model

Then it errors and quits

It looks like an error with gcc version on my end.

```

In file included from /usr/lib/cuda-11.2/bin/../targets/x86_64-linux/include/cuda_runtime.h:83,
                 from <command-line>:
/usr/lib/cuda-11.2/bin/../targets/x86_64-linux/include/crt/host_config.h:139:2: error: #error -- unsupported GNU version! gcc versions later than 10 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
  139 | #error -- unsupported GNU version! gcc versions later than 10 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
      |  ^~~~~
In file included from /usr/lib/cuda-11.2/bin/../targets/x86_64-linux/include/cuda_runtime.h:83,
                 from <command-line>:
/usr/lib/cuda-11.2/bin/../targets/x86_64-linux/include/crt/host_config.h:139:2: error: #error -- unsupported GNU version! gcc versions later than 10 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
  139 | #error -- unsupported GNU version! gcc versions later than 10 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
      |  ^~~~~
In file included from /usr/lib/cuda-11.2/bin/../targets/x86_64-linux/include/cuda_runtime.h:83,
                 from <command-line>:
/usr/lib/cuda-11.2/bin/../targets/x86_64-linux/include/crt/host_config.h:139:2: error: #error -- unsupported GNU version! gcc versions later than 10 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
  139 | #error -- unsupported GNU version! gcc versions later than 10 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
      |  ^~~~~
In file included from /usr/lib/cuda-11.2/bin/../targets/x86_64-linux/include/cuda_runtime.h:83,
                 from <command-line>:
/usr/lib/cuda-11.2/bin/../targets/x86_64-linux/include/crt/host_config.h:139:2: error: #error -- unsupported GNU version! gcc versions later than 10 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
  139 | #error -- unsupported GNU version! gcc versions later than 10 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
      |  ^~~~~
cc1plus: note: unrecognized command-line option ‘-Wno-return-type-c-linkage’ may have been intended to silence earlier diagnostics
cc1plus: note: unrecognized command-line option ‘-Wno-return-type-c-linkage’ may have been intended to silence earlier diagnostics
cc1plus: note: unrecognized command-line option ‘-Wno-return-type-c-linkage’ may have been intended to silence earlier diagnostics
cc1plus: note: unrecognized command-line option ‘-Wno-return-type-c-linkage’ may have been intended to silence earlier diagnostics
Aborted (core dumped)

**Reply by neworderofjamie** [reply]
Yep, CUDA has fairly strict ranges of supported compilers - see https://docs.nvidia.com/cuda/cuda-installation-guide-linux/ for the latest version

**Comment by russelljjarvis** [comment]
Also if I build the docker file I get this:
```
(base) rjjarvis@pop-os:~/git/genn$ sudo docker run --gpus=all -e LOCAL_USER_ID=`id -u $USER` -e LOCAL_GROUP_ID=`id -g $USER` genn
docker: Error response from daemon: could not select device driver "" with capabilities: [[gpu]].


(base) rjjarvis@pop-os:~/git/genn$ nvcc
nvcc fatal   : No input files specified; use option --help for more information
(base) rjjarvis@pop-os:~/git/genn$ which nvcc
/usr/lib/cuda/bin/nvcc

ERRO[0000] error waiting for container: context canceled 
```

**Comment by russelljjarvis** [comment]
I can't compile the cuda samples. So I have to figure out how to downgrade gcc

 make
make[1]: Entering directory '/usr/lib/cuda-11.2/samples/0_Simple/simpleAtomicIntrinsics_nvrtc'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/usr/lib/cuda-11.2/samples/0_Simple/simpleAtomicIntrinsics_nvrtc'
make[1]: Entering directory '/usr/lib/cuda-11.2/samples/0_Simple/vectorAddDrv'
/usr/lib/cuda-11.2/bin/nvcc -ccbin g++ -I../../common/inc -m64 --threads 0 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_52,code=sm_52 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_70,code=sm_70 -gencode arch=compute_75,code=sm_75 -gencode arch=compute_80,code=sm_80 -gencode arch=compute_86,code=sm_86 -gencode arch=compute_86,code=compute_86 -o vectorAdd_kernel64.fatbin -fatbin vectorAdd_kernel.cu
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
In file included from /usr/lib/cuda-11.2/bin/../targets/x86_64-linux/include/cuda_runtime.h:83,
                 from <command-line>:
/usr/lib/cuda-11.2/bin/../targets/x86_64-linux/include/crt/host_config.h:139:2: error: #error -- unsupported GNU version! gcc versions later than 10 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version check; however, using an unsupported host compiler may cause compilation failure or incorrect run time execution. Use at your own risk.
      |  ^~~~~
make[1]: *** [Makefile:381: vectorAdd_kernel64.fatbin] Error 255
make[1]: Leaving directory '/usr/lib/cuda-11.2/samples/0_Simple/vectorAddDrv'
make: *** [Makefile:51: 0_Simple/vectorAddDrv/Makefile.ph_build] Error 2
(base) rjjarvis@pop-os:/usr/lib/cuda-11.2/samples$ make install
make: *** No rule to make target 'install'.  Stop.

**Comment by russelljjarvis** [comment]
I found out I needed to install an a cuda driver extension for containers.
sudo docker run --rm --runtime=nvidia --gpus all nvidia/cuda:11.6.2-base-ubuntu20.04 nvidia-smi

Now returns
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 450.51.06    Driver Version: 450.51.06    CUDA Version: 11.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            On   | 00000000:00:1E.0 Off |                    0 |
| N/A   34C    P8     9W /  70W |      0MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

**Reply by neworderofjamie** [reply]
Yep, as the readme says:

> To work with such CUDA-enabled containers, you need to first install CUDA on your host system as described above and then install docker and the NVIDIA Container Toolkit as described in https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#docker. You can then build the GeNN container yourself or download it from Dockerhub.
